Data/ByteString/Short/Internal.hs

```diff
@@ -482,6 +487,56 @@ toShortIO (BS fptr len) = do
     touchForeignPtr fptr
     ShortByteString <$> stToIO (unsafeFreezeByteArray mba)
 
+-- | A simple wrapper around 'fromShort' that wraps the strict 'ByteString' as
+-- a one-chunk lazy 'LBS.ByteString'.
+lazyFromShort :: ShortByteString -> LBS.ByteString
+lazyFromShort = LBS.fromStrict . fromShort
+
+-- | /O(n)/. Convert a lazy 'LBS.ByteString' into a 'ShortByteString'.
+--
+-- This makes a copy, so does not retain the input string.  Naturally, best
+-- used only with sufficiently short lazy ByteStrings.
+--
+lazyToShort :: LBS.ByteString -> ShortByteString
+lazyToShort LBS.Empty = empty
+lazyToShort (LBS.Chunk c@(BS _ len) cs) =
+    unsafeDupablePerformIO $ do
+        mba0 <- stToIO (newByteArray len)
+        mba <- lazyToShortIO mba0 0 len c cs
+        ShortByteString <$> stToIO (unsafeFreezeByteArray mba)
+
+-- | Append first and any subsequent chunks of lazy bytestring to the provided
+-- buffer, resizing it if required.  The return value is the final buffer.
+-- Arithmetic is done with care, to avoid undetected integer addition
+-- wrap-around.  While nobody should be copying multi-gigabyte lazy bytestrings
+-- to "short" bytestrings, we should error out if possible, rather than corrupt
+-- memory..
+lazyToShortIO :: MutableByteArray RealWorld      -- ^ result buffer
+              -> Int                             -- ^ space already used
+              -> Int                             -- ^ total space allocated
+              -> ByteString                      -- ^ initial strict chunk
+              -> LBS.ByteString                  -- ^ remaining chunks
+              -> IO (MutableByteArray RealWorld) -- ^ possibly resized result
+lazyToShortIO mba used size c@(BS fptr len) cs
+    -- Safe against overflow since used <= size.
+    | len <= size - used
+      = do let ptr = unsafeForeignPtrToPtr fptr
+           stToIO (copyAddrToByteArray ptr mba used len)
+           touchForeignPtr fptr
+           case cs of
+               LBS.Empty
+                   -> stToIO (shrinkByteArray mba $ used + len) >> pure mba
+               LBS.Chunk c' cs'
+                   -> lazyToShortIO mba (used + len) size c' cs'
+    -- Care to detect possible 'Int' arithmetic overflow.
+    | used < used + len
+    , let newsize = max (used + len) (size + size `quot` 2)
```

Comment:
* If ``(size + size `quot` 2)`` overflows and becomes negative, this will behave degenerately and cause O(n²) allocation for inputs that narrowly fit into the `Int#` size limit.
* Signed ``(`div` 2)`` will compile to an arithmetic right shift; signed ``(`quot` 2)`` needs an extra shift and add/subtract to fix up negative odd inputs. (Or ``(`shiftR` 1)`` makes the performance question much easier.)